Duplicati/Library/Main/Operation/DeleteHandler.cs

```diff
@@ -156,222 +158,233 @@ public void DoRun(Database.LocalDeleteDatabase db, ref System.Data.IDbTransactio
                         Logging.Log.WriteDryrunMessage(LOGTAG, "WouldDeleteHelp", "Remove --dry-run to actually delete files");
                 }
                 
-                if (!m_options.NoAutoCompact && (forceCompact || (toDelete != null && toDelete.Length > 0)))
+                if (!m_options.NoAutoCompact && (forceCompact || toDelete.Length > 0))
                 {
                     m_result.CompactResults = new CompactResults(m_result);
                     new CompactHandler(m_backendurl, m_options, (CompactResults)m_result.CompactResults).DoCompact(db, true, ref transaction, sharedManager);
                 }
-                
-                m_result.SetResults(
-                    from n in filesetNumbers
-                    where toDelete.Contains(n.Item2)
-                    select n, 
-                    m_options.Dryrun);
+
+                m_result.SetResults(versionsToDelete.Select(v => new Tuple<long, DateTime>(v.Version, v.Time)), m_options.Dryrun);
             }
         }
+    }
 
-        /// <summary>
-        /// Gets the filesets selected for deletion
-        /// </summary>
-        /// <returns>The filesets to delete</returns>
-        /// <param name="allBackups">The list of backups that can be deleted</param>
-        private DateTime[] GetFilesetsToDelete(Database.LocalDeleteDatabase db, DateTime[] allBackups)
+    public abstract class FilesetRemover
+    {
+        protected readonly Options Options;
+
+        protected FilesetRemover(Options options)
         {
-            if (allBackups.Length == 0)
-            {
-                return allBackups;
-            }
+            this.Options = options;
+        }
 
-            DateTime[] sortedAllBackups = allBackups.OrderByDescending(x => x.ToUniversalTime()).ToArray();
+        public abstract IEnumerable<IListResultFileset> GetFilesetsToDelete(IEnumerable<IListResultFileset> filesets);
+    }
 
-            if (sortedAllBackups.Select(x => x.ToUniversalTime()).Distinct().Count() != sortedAllBackups.Length)
-            {
-                throw new Exception($"List of backup timestamps contains duplicates: {string.Join(", ", sortedAllBackups.Select(x => x.ToString()))}");
-            }
+    /// <summary>
+    /// Remove versions specified by the --version option.
+    /// </summary>
+    public class SpecificVersionsRemover : FilesetRemover
+    {
+        public SpecificVersionsRemover(Options options) : base(options)
+        {
+        }
 
-            List<DateTime> toDelete = new List<DateTime>();
+        public override IEnumerable<IListResultFileset> GetFilesetsToDelete(IEnumerable<IListResultFileset> filesets)
+        {
+            ISet<long> versionsToDelete = new HashSet<long>(this.Options.Version ?? new long[0]);
+            return filesets.Where(x => versionsToDelete.Contains(x.Version));

```

Comment:
This is causing the `PurgeBrokenFilesTest` test to fail.  It appears that the `Fileset` table in a `LocalDeleteDatabase` has a state I did not expect, where the newest `DateTime` has the largest version number (typically, version `0` represents the newest backup).  In the test, the options specify that version `2` should be deleted.  While the new implementation here will select fileset version `2` for deletion, the old implementation would take a `DateTime[]` (ordered descending) and remove the 3rd (index `2`) element.  However, the 3rd element is not fileset version `2` in the database.

Either the "version" in `Options.Version` does not mean what I think it means, or there may be a bug somewhere.

Reply:
It turns out I was building the `IListResultFilesets` incorrectly.  The `Version` should not be the `ID` field of the `Fileset` table.  Rather, it should be the backup version exposed to the user (integers where `0` is the most recent backup version.

Revision 60b4c8479cc9bf7982e8909dc45881a4e3e3202e should fix this.